policy_bundle/iam.sp

```diff
@@ -180,6 +180,58 @@ query "iam_api_key_age_90" {
   EOQ
 }
 
+query "iam_api_key_unrestriced" {
+  sql = <<-EOQ
+    select
+      'https://iam.googleapis.com/v1/projects/' || project || '/apikeys/' || name as resource,  display_name,
+      case
+        when restrictions is null then 'alarm'
+        else 'ok'
+      end as status,
+      display_name || ' ' || uid || ' has either host, apps or API restriction.'
+      as reason
+      ${local.common_dimensions_global_sql}
+    from
+      gcp_apikeys_key;
+  EOQ
+}
+
+query "iam_api_key_restricted_to_apis" {
+  sql = <<-EOQ
+    select
+     'https://iam.googleapis.com/v1/projects/' || project || '/apikeys/' || name as resource,
+    case
+      when restrictions -> 'apiTargets' is null then 'alarm'
+      else 'ok'
+    end as status,
+    case
+      when restrictions -> 'apiTargets' is null then title || ' API key is not restricted to required APIs.'
+      else title || ' API key is restricted to only required APIs.'
+    end as reason
+    ${local.common_dimensions_sql}
+  from
+    gcp_apikeys_key;
+  EOQ
+}
+
+query "iam_api_key_restricted_to_hosts_apps" {
```

Comment:
@saisirishreddy  I guess this should satisfy some more conditions as mentioned below; refer CIS_Google_Cloud_Platform_Foundation_Benchmark_v1.2.0

**From Console:**
```
1. Go to APIs & Services\Credentials using
https://console.cloud.google.com/apis/credentials
2. In the section API Keys, Click the API Key Name. The API Key properties display on a
3. For every API Key, ensure the section Key restrictions parameter Application
restrictions is not set to None.
Or,
Ensure **Application restrictions** is set to HTTP referrers and the referrer is not set to
wild-cards (* or *.[TLD] or *.[TLD]/*) allowing access to any/wide HTTP
referrer(s)
Or,
Ensure **Application restrictions** is set to IP addresses and referrer is not set to any
host (0.0.0.0 or 0.0.0.0/0 or ::0)
```
After testing a few scenarios, I recommend addressing all combinations of scenarios.

For this first condition > `when restrictions -> 'serverKeyRestrictions' is null`, we should check other conditions, including `browserKeyRestrictions`, `androidKeyRestrictions` and `iosKeyRestrictions`

The reason is; the Key restrictions allow to set up multiple options such as HTTP referrers, IP Adresses, Android Apps, iOs Apps; hence only, restrictions -> 'serverKeyRestrictions' only satisfies only Adresses

![image](https://github.com/turbot/steampipe-mod-gcp-compliance/assets/26873346/7e62fc0d-38f3-4f9b-858a-c278c2fa8877)

I am recommending this below query for `iam_api_key_restricted_to_hosts_apps` which satisfies the CIS need of CIS steps, I hope these steps same for all CIS versions (Yet to check)

```sql
select
  'https://iam.googleapis.com/v1/projects/' || project || '/apikeys/' || name as resource,
  case
    when restrictions -> 'serverKeyRestrictions' is null and restrictions -> 'browserKeyRestrictions' is null and restrictions -> 'androidKeyRestrictions' is null and restrictions -> 'iosKeyRestrictions' is null then 'alarm'
    when restrictions -> 'serverKeyRestrictions' @> any( array[
    '{"allowedIps": ["0.0.0.0"]}', '{"allowedIps": ["0.0.0.0/0"]}', '{"allowedIps": ["::0"]}']::jsonb[]) then 'alarm'
    when restrictions -> 'browserKeyRestrictions' @> any( array[
    '{"allowedReferrers": ["*"]}','{"allowedReferrers": ["*.[TLD]/*"]}','{"allowedReferrers": ["*.[TLD]"]}' ]::jsonb[] ) then 'alarm'
  else 'ok'
    end as status,
    case
      when restrictions -> 'serverKeyRestrictions' is null and restrictions -> 'browserKeyRestrictions' is null and restrictions -> 'androidKeyRestrictions' is null and restrictions -> 'iosKeyRestrictions' is null then 'API key not restricted to use any specified Websites, Hosts and Apps.'
      when restrictions -> 'serverKeyRestrictions' @> any( array[
      '{"allowedIps": ["0.0.0.0"]}', '{"allowedIps": ["0.0.0.0/0"]}', '{"allowedIps": ["::0"]}']::jsonb[]) then 'API key open to any hosts.'
      when restrictions -> 'browserKeyRestrictions' @> any( array[
      '{"allowedReferrers": ["*"]}','{"allowedReferrers": ["*.[TLD]/*"]}','{"allowedReferrers": ["*.[TLD]"]}' ]::jsonb[] ) then 'API key open to any or wide range of HTTP referrer(s).'
      else title || ' API key is restricted to use by only specified Hosts and Apps.'
    end as reason
  from
    gcp_apikeys_key;

```

Reply:
This could be the change in the iam.sp

```sql
query "iam_api_key_restricted_to_hosts_apps" {
  sql = <<-EOQ
    select
      'https://iam.googleapis.com/v1/projects/' || project || '/apikeys/' || name as resource,
      case
        when restrictions -> 'serverKeyRestrictions' is null
          and restrictions -> 'browserKeyRestrictions' is null
          and restrictions -> 'androidKeyRestrictions' is null
          and restrictions -> 'iosKeyRestrictions' is null then 'alarm'
        when restrictions -> 'serverKeyRestrictions' @> any( array[
        '{"allowedIps": ["0.0.0.0"]}', '{"allowedIps": ["0.0.0.0/0"]}', '{"allowedIps": ["::0"]}']::jsonb[]) then 'alarm'
        when restrictions -> 'browserKeyRestrictions' @> any( array[
        '{"allowedReferrers": ["*"]}','{"allowedReferrers": ["*.[TLD]/*"]}','{"allowedReferrers": ["*.[TLD]"]}' ]::jsonb[] ) then 'alarm'
        else 'ok'
      end as status,
      case
        when restrictions -> 'serverKeyRestrictions' is null
          and restrictions -> 'browserKeyRestrictions' is null
          and restrictions -> 'androidKeyRestrictions' is null
          and restrictions -> 'iosKeyRestrictions' is null
          then title || ' API key not restricted to use any specified Websites, Hosts and Apps.'
        when restrictions -> 'serverKeyRestrictions' @> any( array[
        '{"allowedIps": ["0.0.0.0"]}', '{"allowedIps": ["0.0.0.0/0"]}', '{"allowedIps": ["::0"]}']::jsonb[]) then title || ' API key open to any hosts.'
        when restrictions -> 'browserKeyRestrictions' @> any( array[
        '{"allowedReferrers": ["*"]}','{"allowedReferrers": ["*.[TLD]/*"]}','{"allowedReferrers": ["*.[TLD]"]}' ]::jsonb[] ) then  title || ' API key open to any or wide range of HTTP referrer(s).'
        else title || ' API key is restricted with specific Website(s), Host(s) and App(s).'
      end as reason
      ${local.common_dimensions_sql}
    from
      gcp_apikeys_key;
  EOQ
}
```

![image](https://github.com/turbot/steampipe-mod-gcp-compliance/assets/26873346/45abf711-5785-44a3-b643-c66a2b50b722)

Reply:
yes, looks good.